autogpt/core/llm/base.py

```diff
@@ -1,5 +1,145 @@
 import abc
+import typing
+from dataclasses import dataclass, field
+from typing import Dict, List
 
+if typing.TYPE_CHECKING:
+    from autogpt.core.configuration.base import Configuration
+    from autogpt.core.logging.base import Logger
+    from autogpt.core.planning.base import ModelPrompt
+    from autogpt.core.plugin.base import PluginManager
+    from autogpt.core.workspace.base import Workspace
+
+
+@dataclass
+class ModelInfo:
+    """Struct for model information.
+
+    Would be lovely to eventually get this directly from APIs, but needs to be scraped from
+    websites for now.
+
+    """
+
+    name: str
+    prompt_token_cost: float
+    completion_token_cost: float
+    max_tokens: int
+
+
+@dataclass
+class ChatModelInfo(ModelInfo):
+    """Struct for chat model information."""
 
-class LanguageModel(abc.ABC):
     pass
+
+
+@dataclass
+class EmbeddingModelInfo(ModelInfo):
+    """Struct for embedding model information."""
+
+    embedding_dimensions: int
+
+
+@dataclass
+class LLMResponse:
+    """Standard response struct for a response from an LLM model."""
+
+    model_info: ModelInfo
+    prompt_tokens_used: int = 0
+    completion_tokens_used: int = 0
+
+
+@dataclass
+class EmbeddingModelResponse(LLMResponse):
+    """Standard response struct for a response from an embedding model."""
+
+    embedding: List[float] = field(default_factory=list)
+
+    def __post_init__(self):
+        if self.completion_tokens_used:
+            raise ValueError("Embeddings should not have completion tokens used.")
+
+
+@dataclass
+class ChatModelResponse(LLMResponse):
+    """Standard response struct for a response from an LLM model."""
+
+    content: str = None
+
+
+class LanguageModel(abc.ABC):
+    configuration_defaults = {"language_model": {}}
+
+    @abc.abstractmethod
+    def __init__(
+        self,
+        configuration: Configuration,
+        logger: Logger,
+        workspace: Workspace,
+        plugin_manager: PluginManager,
+    ):
+        ...
+
+    @abc.abstractmethod
+    def list_models(self) -> Dict[str, ModelInfo]:
+        """List all available models."""
+        ...
+
+    @abc.abstractmethod
+    def get_model_info(self, model_name: str) -> ModelInfo:
+        """Get information about a specific model."""
```

Comment:
Where do we add new models?

Reply:
That's a good question and one I'll show in concrete implementation.  Essentially the core system will define a couple of models.  The plugin system may register new ones (eventually).  Part of the language model definition will include concrete instances of `LanguageModelInfo` and `EmbeddingModelInfo` objects.